File: src/cpp/integrate.cpp
#include "integrate.h"

void leapfrog_start(sim_state_t* s, double dt){
        const float* a = s->a;
        float*  vh = s->vh;
        float*  v = s->v;
        float*  x = s->x;
        int n = s->n;

        for (int i = 0; i < 2*n; ++i) vh[i] = v[i] + a[i] * dt/2;
        for (int i = 0; i < 2*n; ++i) v[i] += a[i] * dt;
        for (int i = 0; i < 2*n; ++i) x[i] += vh[i] * dt;

        reflect_bc(s);
}

void leapfrog_step(sim_state_t* s, double dt){
        const float*  a = s->a;
        float*  vh = s->vh;
        float*  v = s->v;
        float*  x = s->x;
        int n = s->n;

        for (int i = 0; i < 2*n; ++i) vh[i] += a[i] * dt;
        for (int i = 0; i < 2*n; ++i) v[i] = vh[i] + a[i] * dt / 2;
        for (int i = 0; i < 2*n; ++i) x[i] += vh[i] * dt;

        reflect_bc(s);
}

File: src/integrate.f90
module integrate

  ! Implementation of functions and subroutines for the numerical integration
  ! current method: leapfrog method

  !Created by Jannik Zuern on 05/16/2016
  !Last modified: 07/05/2016

  implicit none


  contains

  subroutine leapfrog_start(state,params)

    use util
    use sphfunctions

    type(systemstate)            :: state
    type(sim_parameter)					 :: params
    double precision             :: dt
    integer                      :: startidx, endidx

    ! only update liquid particle positions according to integration scheme
    startidx = 3*state%nSolidParticles+1
    endidx   = 3*state%nParticles

    dt = params%dt

    state%vh(startidx:endidx)  = state%v(startidx:endidx)  + (state%a (startidx:endidx)   * dt/2)
    state%v (startidx:endidx)  = state%v(startidx:endidx)  + (state%a (startidx:endidx)   * dt  )
    state%x (startidx:endidx)  = state%x(startidx:endidx)  + (state%vh(startidx:endidx)   * dt  )

    ! if we have a water mill, update the positions of the solid particles as well
    if (params%mill) then
      call update_solid_particles_positions(state,params)
    end if

    ! particles located at the wall will be reflected
    call reflect_bc(state)

    return
  end


  subroutine leapfrog_step(state,params)
    use util
    use sphfunctions

    type (systemstate)             :: state
    type(sim_parameter)						 :: params
    double precision               :: dt
    integer                        :: startidx,endidx

    ! only update liquid particle positions according to integration scheme
    startidx = 3*state%nSolidParticles+1
    endidx   = 3*state%nParticles

    dt = params%dt

    state%vh(startidx:endidx) = state%vh(startidx:endidx) + state%a(startidx:endidx) * dt
    state%v(startidx:endidx)  = state%vh(startidx:endidx) + state%a(startidx:endidx) * dt/2
    state%x(startidx:endidx)  = state%x(startidx:endidx)  + state%vh(startidx:endidx)* dt

    ! if we have a water mill, update the positions of the solid particles as well
    if (params%mill) then
      call update_solid_particles_positions(state,params)
    end if

    ! particles located at the wall will be reflected
    call reflect_bc(state)

    return
  end


end module integrate
